src/x11/x.c

```diff
@@ -1124,22 +465,41 @@ static void x_set_wm(Window win)
                 PropModeReplace, (unsigned char *) data, 1L);
 }
 
+GSource* x_win_reg_source(window_x11 *win)
+{
+        static GSourceFuncs x_src_fn = {
+                x_mainloop_fd_prepare,
+                x_mainloop_fd_check,
+                x_mainloop_fd_dispatch,
+                NULL,
+                NULL,
+                NULL
+        };
+
+        struct x11_source *x_src = (struct x11_source*) g_source_new(&x_src_fn,
```

Comment:
err, at least from curiosity: How is this supposed to work? The structure pointer, which gets spit out by `g_source_new` is [opaque](https://developer.gnome.org/glib/stable/glib-The-Main-Event-Loop.html#GSource). This may work, but I actually think this is a very bad hack [like it's done in `dunstify`](https://github.com/dunst-project/dunst/blob/7a9187728c8fbd519358e56e3df55a9639025f15/dunstify.c#L178-L188).

Reply:
> But why are we casting it then again to a GSource* back?!?

We could leave it as-is but we're casting it back because all glib methods need the GSource and not the derived struct so it's just easier that way.

[Apparently this is a really weird way to do inheritance in C.](https://stackoverflow.com/questions/1114349/struct-inheritance-in-c)